### PR TITLE
Add rock for net-istio-controller

### PR DESCRIPTION
**Pull request by mvlassis**
Closes https://github.com/canonical/knative-operators/issues/238

This rock is based on the image found here: https://console.cloud.google.com/gcr/images/knative-releases/global/knative.dev/net-istio/cmd/controller

To check that everything is as expected:
```
# Clone the repository and checkout to this branch, then cd to the proper subdirectory
cd net-istio-controller

# Build the rock
tox -e pack

# Export to docker
tox -e export-to-docker

# Sanity tests
tox -e sanity

# Compare output with the upstream image

docker run gcr.io/knative-releases/knative.dev/net-istio/cmd/controller:v1.12.3
## Sample output
## Output
2024/11/12 11:27:01 Error building kubeconfig: failed to create client config: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable

docker run net-istio-webhook:v1.12.3
# Get the ID of the container
docker ps
# Run bash into it
docker exec -ti <container_id> bash
# Execute the service command manually
/ko-data/controller
## Sample output
# Output
2024/11/12 11:29:29 Error building kubeconfig: failed to create client config: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable

### CODE REVIEW

File: net-istio-controller/rockcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      (echo "# os-release" && cat /etc/os-release && echo "# dpkg-query" && \
+      dpkg-query --root=${CRAFT_PROJECT_DIR}/../bundles/ubuntu-22.04/rootfs/ -f '${db:Status-Abbrev},${binary:Package},${Version},${source:Package},${Source:Version}\n' -W) \
+      > ${CRAFT_PART_INSTALL}/usr/share/rocks/dpkg.query
+
+  net-istio-controller:
+    plugin: go
+    build-snaps:
+      - go/1.21/stable
+    source: https://github.com/knative-extensions/net-istio.git
+    source-tag: release-1.12
```
Comment (NohaIhab) [inline net-istio-controller/rockcraft.yaml:36]: I prefer to use the tag `knative-v1.12.3` from the specific minor release over this branch, to keep consistency of the built rock in case a new minor version was introduced

File: net-istio-controller/rockcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  
+  non-root-user:
+    plugin: nil
+    after: [ net-istio-controller ]
+    overlay-script: |
+      # Create a user in the $CRAFT_OVERLAY chroot
+      groupadd -R $CRAFT_OVERLAY -g 1001 ubuntu
+      useradd -R $CRAFT_OVERLAY -M -r -u 1001 -g ubuntu ubuntu
+    override-prime: |
+      craftctl default
```
Comment (NohaIhab) [inline net-istio-controller/rockcraft.yaml:53]: why is this needed? doesn't the prime step do this by default?

File: net-istio-controller/tests/test_rock.py
Comment (NohaIhab) [inline net-istio-controller/tests/test_rock.py:1]: please see an example of the new structure of the test file [here](https://github.com/canonical/kserve-rocks/blob/main/agent/tests/test_rock.py).

**Review by NohaIhab — COMMENTED** [review]
thank you @mvlassis , left you some comments for the code and will test in parallel as you address them.

File: net-istio-controller/rockcraft.yaml (hunk 2 above)
Comment (mvlassis) [reply net-istio-controller/rockcraft.yaml:53]: I wasn't sure about this. After checking the documentation for `craftctl`, it seems that this step is unnecessary, since we do not perform any additional steps during the `PRIME` step. Updating this!

Comment (mvlassis) [reply net-istio-controller/rockcraft.yaml:53]: Relevant documentation: https://snapcraft.io/docs/using-craftctl#heading--run-default

File: net-istio-controller/rockcraft.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+parts:
+  security-team-requirement:
+    plugin: nil
+    override-build: |
+      mkdir -p ${CRAFT_PART_INSTALL}/usr/share/rocks
+      (echo "# os-release" && cat /etc/os-release && echo "# dpkg-query" && \
+      dpkg-query --root=${CRAFT_PROJECT_DIR}/../bundles/ubuntu-22.04/rootfs/ -f '${db:Status-Abbrev},${binary:Package},${Version},${source:Package},${Source:Version}\n' -W) \
+      > ${CRAFT_PART_INSTALL}/usr/share/rocks/dpkg.query
+
+  net-istio-controller:
```
Comment (NohaIhab) [inline net-istio-controller/rockcraft.yaml:31]: we're missing the packages installed in the base image, please add them to `stage-packages`. You can refer to the approved [spec packages section](https://docs.google.com/document/d/1l_CrR_5U6NmDXpcoH2hkhVuoDzdpwFto6EwsylQ5vRw/edit?tab=t.0#heading=h.wzidaph2gx9s) on how to find the packages.

File: net-istio-controller/rockcraft.yaml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    source: https://github.com/knative-extensions/net-istio.git
+    source-type: git
+    source-tag: knative-v1.12.3
+    build-environment:
+      - CGO_ENABLED: 0
+      - GOOS: linux
+    override-build: |
+      cd cmd/controller
+      mkdir $CRAFT_PART_INSTALL/ko-app
+      go build -o $CRAFT_PART_INSTALL/ko-app/controller -a .
```
Comment (NohaIhab) [inline net-istio-controller/rockcraft.yaml:44]: need to copy the static assets in [kodata](https://github.com/knative-extensions/net-istio/tree/release-1.12/cmd/controller/kodata), refer to [this section of the spec](https://docs.google.com/document/d/1l_CrR_5U6NmDXpcoH2hkhVuoDzdpwFto6EwsylQ5vRw/edit?tab=t.0#heading=h.okpop7e4nq2h)

Comment (NohaIhab) [reply net-istio-controller/rockcraft.yaml:44]: also see #6 for a similar implementation

File: net-istio-controller/tests/test_rock.py
Comment (NohaIhab) [inline net-istio-controller/tests/test_rock.py:1]: when you copy the static assets, please add a sanity test to check they exist in the Rock.

**Review by NohaIhab — CHANGES_REQUESTED** [review]
some additional comments based on the ko spec

**Review by NohaIhab — APPROVED** [review]
LGTM, thank you @mvlassis !

**Review by NohaIhab — APPROVED** [review]
